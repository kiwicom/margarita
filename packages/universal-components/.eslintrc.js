const OFF = 0;
const WARNING = 1;
const ERROR = 2;

module.exports = {
  root: true,
  extends: ['@kiwicom/eslint-config'],
  parser: 'babel-eslint',
  env: {
    es6: true,
    jest: true,
    node: true,
    browser: true,
  },
  rules: {
    'react/sort-comp': ERROR,
    'no-unused-vars': [ERROR, { args: 'after-used' }],
    'no-restricted-imports': [
      ERROR,
      {
        paths: [
          {
            name: 'react-native',
            importNames: ['StyleSheet'],
            message: 'Please use PlatformStyleSheet instead',
          },
        ],
      },
    ],
  },
};
