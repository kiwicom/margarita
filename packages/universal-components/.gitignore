# Storybook
storybook-static

# Docz.site - documentation generating tool
**/.docz/*

#build
lib/