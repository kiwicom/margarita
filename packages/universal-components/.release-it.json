{
  "increment": "conventional:angular",
  "scripts": {
    "beforeStage": "yarn build"
  },
  "git": {
    "commitMessage": "chore: release %s",
    "tagName": "v%s"
  },
  "npm": {
    "publish": true
  },
  "github": {
    "release": true
  },
  "access": "public"
}