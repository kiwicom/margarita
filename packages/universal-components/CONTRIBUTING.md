# Contributing Guidelines

_Note: These guidelines assume you have an existing React Native setup on your machine and that your simulator/emulator is properly functionning. If not, have a look at [Getting Started â€¢ React Native](https://facebook.github.io/react-native/docs/getting-started.html), and go to "Building Projects with Native Code"._

To run the storybook and start developing, clone the repository

```bash
git clone https://github.com/kiwicom/universal-components.git
```

and enter the directory to install the node modules:

```bash
cd universal-components
yarn
```

## Desktop

If you want to see the web version, run

```bash
yarn storybook
```

and you should go to <http://localhost:9009> to see

![](./docs/assets/storybook-web.png)

## Mobile

After running `yarn`, you also need to link assets using:

```bash
react-native link
```

### iOS

To run the storybook on iOS, execute

```bash
yarn ios
```

to see ![](./docs/assets/storybook-ios.png)

### Android

To run the storybook on Android, execute

```bash
yarn android
```

### Easier navigation

Although it is possible to navigate using the menu inside the mobile storybook, there is a navigator inside the browser available at <http://localhost:7007> after you run

```bash
yarn storybook-native
```

You should see something similar to ![](./docs/assets/storybook-native.png)

and it is then easier to test the interactions with the action logger, and play with the knobs inside the Playground.

## Troubleshooting

- **I added `MyComponent.stories.js` and it is not available on iOS or Android in the storybook, but I can see it on the web version.**<br>
  This is sometimes due to the fact that `react-native-story-loader` did not run. To fix it, execute

  ```bash
  yarn rnstl
  ```

- **I started the web UI for the mobile storybook but no stories are displayed.**<br>
  Make sure that the packager is running (`yarn ios` or `yarn android`), and refresh the application by pressing `CMD+R` on iOS or by pressing `R` twice on `Android`.

## Visual regression testing

We use [Loki](https://loki.js.org) to perform visual testing; this ensures that the appearance of components does not change without us noticing.

### Usage

- on Chrome, you need to have the Storybook started (`yarn storybook`) and then run `yarn loki:test:chrome`. If any error is reported, you can check the difference in `.loki/difference`. If you expected these changes, run `yarn loki:approve:chrome` to update the screenshots.
- on iPhone, you need to have the Storybook started (`yarn storybook-native`) and the app open on the simulator (`yarn native`). Run `yarn loki:test:iphone`. If any error is reported, you can check the difference in `.loki/difference`. If you expected these changes, run `yarn loki:approve:iphone` to update the screenshots.

### Skipping context dependent stories

If a story is too dependent on context and the output changes on every render (e.g. the Playground story for the DatePicker component: a new date is generated on every render), you can use `.lokiSkip` instead of `.add` to add your story to Storybook. That story would then be visible in the Storybook and be ignored by Loki.

## Before submitting a PR: a checklist

- I tested my component at least on iOS or on Android;
- I wrote tests for my component and all tests are passing;
- I ran `yarn test-ci` and there were no errors;
- I ran `yarn loki:test:chrome` and `yarn loki:test:iphone` and there were no errors;
- I squashed all my commits into one (unless having several really makes sense);
- I chose a descriptive message for all my commit messages ([check commit message convention](#commit-message-convention));
- I created my PR and added reviewers.

### Commit message convention

- Commit message should contain also proper description, once you open a PR it will be added automatically to PR description.

- We prefix our commit messages with one of the following to signify the kind of change:

  - `Fix`: bug fixes, e.g. fix some colors, paddings.
  - `Feat`: new features, e.g. add new component.
  - `Refactor`: code/structure refactor, e.g. new structure folder for components/ extracting component to separate file.
  - `Docs`: changes into documentation,
  - `Test`: adding or updating tests, e.g. unit, snapshot testing.
  - `Chore`: tooling changes, e.g. change circle ci config.
  - `BREAKING`: for changes that break existing usage, e.g. change API of a component.

**To make this easier**, please run the following command:

```bash
git config --local commit.template docs/commit-template.txt
```

This will pre-fill your commit message with a [template](docs/commit-template.txt) whenever you run `git commit`, reminding you to follow the convention and to add a summary.

## Updating Icons

If you need to update the icon list (`Icon/icons.json`), make sure to run

```bash
node scripts/generateTypes.js
```

to update `src/types/_generated-types/index.js`. This ensures flow types are accurate.

## Style Guide

### Writing flow types in different files

When writing `flow` types, leave them as close to the component as possible, ideally in the same file. They should be written just after the imports and before the component declaration.

However, since this is a multi-platform component library, if some types need to be reused in different files specific to each platform, the accepted practice is to extract said types into a file `ComponentTypes.js` in the same `Component` folder. They would typically consist of `Props` and `State` types.

> For instance, have a look at [Checkbox](./src/Checkbox). Since we have a `Checkbox.native.js` file and a `Checkbox.web.js` file, we need to extract the common types into `CheckboxTypes.js`. However, `CheckboxShared.js` is a cross-platform component and it contains its types inside the same file.

If you need to import types in other components, the convention is to do the following:

```javascript
import type { Props as ComponentProps } from '../Component/Component';
```
