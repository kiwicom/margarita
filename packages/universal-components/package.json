{
  "name": "@kiwicom/universal-components",
  "version": "0.0.14",
  "description": "Component Library for React Native, Expo and React applications",
  "homepage": "https://github.com/kiwicom/universal-components",
  "main": "src/index.js",
  "scripts": {
    "storybook": "start-storybook -p 9009 -c storybook -s ./fonts",
    "storybook:build": "rimraf storybook-static; build-storybook -c storybook -s ./fonts",
    "storybook:mobile": "yarn workspace @kiwicom/margarita-mobile-storybook start:universal-components",
    "prerelease": "node scripts/preparePackageJson.js",
    "release": "release-it",
    "postrelease": "node scripts/cleanPackageJson.js",
    "build": "rimraf lib && yarn build:generate-icon-types && yarn build:copy:web && yarn build:copy:native && yarn build:copy:flow && yarn build:copy:fonts && yarn build:lib",
    "build:generate-icon-types": "node ./scripts/generateTypes.js",
    "build:copy:web": "copyfiles -u 1 './src/**/*.js' './src/**/*.json' './src/**/assets/**' -e './src/**/*.stories.js' -e './src/**/__tests__/**' -e './src/**/*.native.js' -e './src/**/*.ios.js' -e './src/**/*.android.js' ./lib/web",
    "build:copy:native": "copyfiles -u 1 './src/**/*.js' './src/**/*.json' './src/**/assets/**' -e './src/**/*.stories.js' -e './src/**/__tests__/**' -e './src/**/**.web.js' ./lib/native",
    "build:copy:flow": "flow-copy-source -i '**/*.stories.js' -i '**/*.test.js' -i '**/*.native.js' -i '**/*.android.js' -i '**/*.ios.js' ./src ./lib/web",
    "build:copy:fonts": "copyfiles -u 0 './fonts/*.ttf' './fonts/**/*.ttf' ./lib",
    "build:lib": "babel --no-babelrc --out-dir ./lib/web ./lib/web --plugins=babel-plugin-react-native-web,@babel/plugin-proposal-class-properties,@babel/plugin-proposal-nullish-coalescing-operator,@babel/plugin-proposal-optional-chaining --presets=@babel/react,@babel/flow --compact true"
  },
  "files": [
    "lib"
  ],
  "keywords": [
    "react",
    "native",
    "react-native-web"
  ],
  "author": "kiwi.com",
  "license": "MIT",
  "dependencies": {
    "@kiwicom/orbit-design-tokens": "^0.3.0",
    "@storybook/react-native": "^4.1.13",
    "react-native-modal": "^7.0.2",
    "react-native-multi-slider": "https://github.com/kiwicom/react-native-multi-slider.git#beb71bf5eb210fed393cb29b18032105b725611e",
    "react-native-status-bar-height": "^2.2.0",
    "styled-components": "^4.1.3"
  },
  "devDependencies": {
    "@babel/cli": "^7.2.0",
    "@babel/core": "^7.2.0",
    "@babel/node": "^7.2.0",
    "@babel/plugin-proposal-nullish-coalescing-operator": "^7.2.0",
    "@babel/plugin-proposal-optional-chaining": "^7.2.0",
    "@babel/preset-flow": "^7.0.0",
    "@babel/preset-react": "^7.0.0",
    "@storybook/addon-actions": "^4.1.13",
    "@storybook/addon-knobs": "^4.1.13",
    "@storybook/addon-links": "^4.1.13",
    "@storybook/addon-ondevice-knobs": "^4.1.13",
    "@storybook/react": "^4.1.13",
    "babel-plugin-react-native-web": "^0.10.1",
    "copyfiles": "^2.1.0",
    "flow-copy-source": "^2.0.2",
    "metro-react-native-babel-preset": "^0.50.0",
    "react-native-web-image-loader": "^0.0.7",
    "release-it": "^10.3.1",
    "rimraf": "^2.6.2",
    "snapshot-diff": "^0.5.0"
  },
  "peerDependencies": {
    "@babel/plugin-proposal-optional-chaining": "^7.2.0",
    "@kiwicom/orbit-design-tokens": "^",
    "@storybook/addon-actions": "^",
    "@storybook/addon-knobs": "^",
    "@storybook/addon-links": "^",
    "@storybook/addon-ondevice-knobs": "^",
    "@storybook/react": "^",
    "babel-preset-expo": "^",
    "react": "^",
    "react-dom": "^",
    "react-native": "^",
    "react-native-testing-library": "^",
    "react-test-renderer": "^",
    "styled-components": "^"
  },
  "rnpm": {
    "assets": [
      "lib/fonts"
    ]
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org/"
  },
  "resolutions": {
    "eslint-plugin-jest": "^22.2.1"
  }
}
