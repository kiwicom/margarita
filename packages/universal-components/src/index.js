// @flow

/* Components */
export { Accordion } from './Accordion';
export { AdaptableBadge } from './AdaptableBadge';
export { Badge } from './Badge';
export { Button } from './Button';
export { Card } from './Card';
export { CarrierLogo } from './CarrierLogo';
export { Checkbox } from './Checkbox';
export { FilterButton } from './FilterButton';
export { FormLabel } from './FormLabel';
export { Icon } from './Icon';
export { Loader, PageLoader } from './Loader';
export { Notification } from './Notification';
export { LocalizedPrice } from './LocalizedPrice';
export { RadioButton } from './RadioButton';
export { Rating } from './Rating';
export { ServiceLogo } from './ServiceLogo';
export { Stepper } from './Stepper';
export { Text } from './Text';
export { TextInput } from './TextInput';
export { Tooltip, TooltipBubble } from './Tooltip';
export { DatePicker } from './DatePicker';
export { Hoverable } from './Hoverable';
export { Slider } from './Slider';
export { Touchable } from './Touchable';
export { SegmentedButton } from './SegmentedButton';
export { Modal } from './Modal';
export { TagsInput } from './TagsInput';
export { OptionPicker } from './OptionPicker';
export {
  MenuGroup,
  MenuItem,
  MenuItemWrapper,
  MenuDescription,
} from './MenuGroup';
export { Picker } from './Picker';
export { PickerButton } from './PickerButton';

/* Utils */
export { StyleSheet } from './PlatformStyleSheet';
export { ThemeProvider, withTheme } from './ThemeProvider';

/* Types */
export type { IconNameType } from './types/_generated-types';
export type { Theme } from './ThemeProvider';

export type {
  StylePropType,
  StyleObjectType,
  PlatformStyleObjectType,
} from './PlatformStyleSheet';

/* Fonts */

import Roboto from '../fonts/Roboto/Roboto-Regular.ttf';
import RobotoItalic from '../fonts/Roboto/Roboto-Italic.ttf';
import RobotoBold from '../fonts/Roboto/Roboto-Bold.ttf';
import RobotoBoldItalic from '../fonts/Roboto/Roboto-BoldItalic.ttf';
import OrbitIcons from '../fonts/orbit-icons.ttf';

const Fonts = {
  Roboto,
  RobotoItalic,
  RobotoBold,
  RobotoBoldItalic,
  OrbitIcons,
};

export { Fonts };
