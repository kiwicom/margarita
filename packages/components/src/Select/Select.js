// @flow

import * as React from 'react';
import { View } from 'react-native';
import { StyleSheet, Button } from '@kiwicom/universal-components';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import SelectOption from './SelectOption';

type Props = {|
  +optionsData: {
    [string]: {|
      +icon: string,
      +label: string,
    |},
  },
  +selectedType: string,
  +onSelect: (id: string) => void,
  +onClosePress: () => void,
|};

const Select = ({
  optionsData,
  selectedType,
  onSelect,
  onClosePress,
}: Props) => {
  const optionKeys = Object.keys(optionsData);
  const options = optionKeys.map((key, index) => {
    const optionData = optionsData[key];
    return (
      <SelectOption
        key={key}
        id={key}
        icon={optionData.icon}
        label={optionData.label}
        selected={selectedType === key}
        underline={index < optionKeys.length - 1}
        onSelect={onSelect}
      />
    );
  });
  return (
    <>
      {options}
      <View style={styles.buttonWrapper}>
        <Button onPress={onClosePress} type="secondary">
          Close
        </Button>
      </View>
    </>
  );
};

const styles = StyleSheet.create({
  buttonWrapper: {
    margin: parseInt(defaultTokens.spaceXSmall, 10),
  },
});
export default Select;
