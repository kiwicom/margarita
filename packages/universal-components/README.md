# Universal Components

Component library compatible with Expo/React Native projects as well as React applications.

Relies on [react-native-web](https://github.com/necolas/react-native-web) to port components, written with a React Native first approach, to the web.

[ðŸ“˜ Storybook](https://kiwicom-universal-components.netlify.com) |
[ðŸ“š Documentation](https://kiwicom-universal-components-docs.netlify.com)

## Usage

```bash
yarn install @kiwicom/universal-components
# Only for React Native
react-native link @kiwicom/universal-components
```

### In Expo projects

Since `react-native link` is not available in Expo, we need to load the fonts at the root of the application. Your `App.js` should look like

```javascript
import React from "react";
import { View } from "react-native";
import { Icon, Text, PageLoader } from "@kiwicom/universal-components";
import { Font } from "expo";
import OrbitIcons from "@kiwicom/universal-components/lib/fonts/orbit-icons.ttf";
import Roboto from "@kiwicom/universal-components/lib/fonts/Roboto/Roboto-Regular.ttf";
import RobotoItalic from '@kiwicom/universal-components/lib/fonts/Roboto/Roboto-Italic.ttf';
import RobotoBold from '@kiwicom/universal-components/lib/fonts/Roboto/Roboto-Bold.ttf';
import RobotoBoldItalic from '@kiwicom/universal-components/lib/fonts/Roboto/Roboto-BoldItalic.ttf';

export default class App extends React.Component {
  state = {
    fontLoaded: false
  };

  componentDidMount() {
    Font.loadAsync({
      "orbit-icons": OrbitIcons,
      Roboto,
      RobotoItalic,
      RobotoBold,
      RobotoBoldItalic,
    }).then(() => {
      this.setState({ fontLoaded: true });
    });
  }

  render() {
    return this.state.fontLoaded ? (
      <View>
        <Text>Open up App.js to start working on your app!</Text>
        <Icon name="calendar" />
      </View>
    ) : (
      <PageLoader />
    );
  }
}
```

For more information, consult the documentation on [expo.io](https://docs.expo.io/versions/latest/guides/using-custom-fonts).

Note: For web projects, you need to ensure you support the `.web.js` extension. [create-react-app](https://github.com/facebook/create-react-app/blob/6364bbf6dc8244508398f934d0882f05e0cb5dcc/packages/react-scripts/config/paths.js#L52) already supports it by default.

# Contributing

See [CONTRIBUTING.md](https://github.com/kiwicom/universal-components/blob/master/CONTRIBUTING.md)

Repository: <https://github.com/kiwicom/universal-components>

Forward your suggestions, issues and bugs [here](https://github.com/kiwicom/universal-components/issues).
