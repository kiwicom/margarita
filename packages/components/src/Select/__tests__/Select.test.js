// @flow

import * as React from 'react';
import { shallow, render, fireEvent } from 'react-native-testing-library';
import { Button } from '@kiwicom/universal-components';

import Select from '../Select';

const optionsData = {
  oneWay: {
    icon: 'flight-direct',
    label: 'One way',
  },
  return: {
    icon: 'flight-return',
    label: 'Return',
  },
};
const onSelect = jest.fn();
const onClose = jest.fn();

const { getByType, getByText } = render(
  <Select
    optionsData={optionsData}
    selectedType={Object.keys(optionsData)[0]}
    onSelect={onSelect}
    onClosePress={onClose}
  />,
);

it('renders', () => {
  expect(
    shallow(
      <Select
        optionsData={optionsData}
        selectedType={Object.keys(optionsData)[0]}
        onSelect={jest.fn()}
        onClosePress={jest.fn()}
      />,
    ),
  ).toMatchSnapshot();
});

it('should execute onSelect method with proper value', () => {
  const testKey = Object.keys(optionsData)[1];
  fireEvent(getByText(optionsData[testKey].label), 'press');
  expect(onSelect).toHaveBeenCalledWith(testKey);
});

it('should execute onClosePress method', () => {
  fireEvent(getByType(Button), 'press');
});
