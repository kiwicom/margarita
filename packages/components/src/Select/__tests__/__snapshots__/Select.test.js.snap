// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
Object {
  "output": <React.Fragment>
    <SelectOption
      icon="flight-direct"
      id="oneWay"
      label="One way"
      onSelect={[MockFunction]}
      selected={true}
      underline={true}
    />
    <SelectOption
      icon="flight-return"
      id="return"
      label="Return"
      onSelect={[MockFunction]}
      selected={false}
      underline={false}
    />
    <Component
      style={
        Object {
          "margin": 8,
        }
      }
    >
      <Button
        onPress={[MockFunction]}
        type="secondary"
      >
        Close
      </Button>
    </Component>
  </React.Fragment>,
}
`;
